lib/algolia.ex

```diff
@@ -196,113 +196,108 @@ defmodule Algolia do
     |> Path.join(path)
   end
 
-  defp request_headers() do
-    [
-      "X-Algolia-API-Key": api_key(),
-      "X-Algolia-Application-Id": application_id()
+  defp request_headers(request_options) do
+    custom = request_options[:headers] || []
+
+    default = [
+      {"X-Algolia-API-Key", api_key()},
+      {"X-Algolia-Application-Id", application_id()}
     ]
+
+    custom ++ default
   end
 
   @doc """
   Get an object in an index by objectID
   """
-  def get_object(index, object_id) do
+  def get_object(index, object_id, opts \\ []) do
     path = "#{index}/#{object_id}"
 
     :read
-    |> send_request(%{method: :get, path: path})
+    |> send_request(%{method: :get, path: path, options: opts[:request_options]})
     |> inject_index_into_response(index)
   end
 
   @doc """
   Add an Object
-  """
-  def add_object(index, object) do
-    body = Jason.encode!(object)
-    path = "#{index}"
-
-    :write
-    |> send_request(%{method: :post, path: path, body: body})
-    |> inject_index_into_response(index)
-  end
 
-  @doc """
-  Add an object with an attribute as the objectID
+  An attribute can be chosen as the objectID.
   """
-  def add_object(index, object, id_attribute: id_attribute) do
-    save_object(index, object, id_attribute: id_attribute)
+  def add_object(index, object, opts \\ []) do
+    if opts[:id_attribute] do
+      save_object(index, object, opts)
+    else
+      body = Jason.encode!(object)
+      path = "#{index}"
+
+      :write
+      |> send_request(%{method: :post, path: path, body: body, options: opts[:request_options]})
+      |> inject_index_into_response(index)
+    end
   end
 
   @doc """
   Add multiple objects
-  """
-  def add_objects(index, objects) do
-    objects
-    |> build_batch_request("addObject")
-    |> send_batch_request(index)
-  end
 
-  @doc """
-  Add multiple objects, with an attribute as objectID
+  An attribute can be chosen as the objectID.
   """
-  def add_objects(index, objects, id_attribute: id_attribute) do
-    save_objects(index, objects, id_attribute: id_attribute)
+  def add_objects(index, objects, opts \\ []) do
+    if opts[:id_attribute] do
+      save_objects(index, objects, opts)
+    else
+      objects
+      |> build_batch_request("addObject")
+      |> send_batch_request(index, opts[:request_options])
+    end
   end
 
   @doc """
   Save a single object, without objectID specified, must have objectID as
   a field
   """
-  def save_object(index, object, id_attribute: id_attribute) do
-    object_id = object[id_attribute] || object[to_string(id_attribute)]
-
-    if !object_id do
-      raise ArgumentError,
-        message: "Your object #{object} does not have a attribute #{id_attribute}"
-    end
+  def save_object(index, object, opts \\ [])
 
-    save_object(index, object, object_id)
+  def save_object(index, object, id) when is_map(object) and not is_list(id) do
```

Comment:
We should consider deprecating this format, since it is not available in the official clients.